File: skills/optimization/reference/src/l-bfgs.ts
/**
 * L-BFGS (Limited-memory BFGS) optimization.
 *
 * Uses the two-loop recursion to implicitly compute H*g without storing the
 * full n×n inverse Hessian matrix. Stores only the last m correction pairs
 * (s_k, y_k), making it suitable for large-scale problems.
 *
 * @node l-bfgs
 * @depends-on vec-ops, result-types, line-search, finite-diff
 * @contract l-bfgs.test.ts
 * @hint memory: m=10 is the default across scipy, Optim.jl, and most implementations.
 * @hint two-loop: The two-loop recursion (Nocedal & Wright Algorithm 7.4) is the
 *       core of L-BFGS. It computes H*g in O(mn) time and O(mn) space.
 * @hint initial-scaling: H0 = (yk'*sk)/(yk'*yk) * I is the standard scaling
 *       from Nocedal & Wright Eq. 7.20. This adapts the initial Hessian
 *       approximation to the local curvature.
 */

import { dot, sub, negate, normInf, addScaled, scale } from "./vec-ops";
import {
  type OptimizeResult,
  type OptimizeOptions,
  defaultOptions,
  checkConvergence,
  isConverged,
  convergenceMessage,
} from "./result-types";
import { wolfeLineSearch } from "./line-search";
import { forwardDiffGradient } from "./finite-diff";

export interface LBFGSOptions extends OptimizeOptions {
  /** Number of correction pairs to store. Default: 10 */
  memory: number;
}

export function defaultLBFGSOptions(
  overrides?: Partial<LBFGSOptions>,
): LBFGSOptions {
  return {
    ...defaultOptions(overrides),
    memory: 10,
    ...overrides,
  };
}

/**
 * Two-loop recursion for L-BFGS search direction.
 *
 * Computes H * g implicitly using the stored correction pairs.
 *
 * @provenance Nocedal & Wright, Algorithm 7.4
 */
function twoLoopRecursion(
  g: number[],
  sHistory: number[][],
  yHistory: number[][],
  rhoHistory: number[],
  gamma: number,
): number[] {
  const m = sHistory.length;
  const n = g.length;

  let q = g.slice();
  const alphas = new Array(m);

  // First loop: from most recent to oldest
  for (let i = m - 1; i >= 0; i--) {
    alphas[i] = rhoHistory[i] * dot(sHistory[i], q);
    q = addScaled(q, yHistory[i], -alphas[i]);
  }

  // Apply initial Hessian approximation: r = gamma * I * q
  let r = scale(q, gamma);

  // Second loop: from oldest to most recent
  for (let i = 0; i < m; i++) {
    const beta = rhoHistory[i] * dot(yHistory[i], r);
    r = addScaled(r, sHistory[i], alphas[i] - beta);
  }

  return r;
}

/**
 * Minimize a function using L-BFGS.
 *
 * @provenance Algorithm from Nocedal & Wright, Chapter 7
 * @provenance Memory m=10 matches scipy (L-BFGS-B maxcor=10) and Optim.jl (m=10)
 * @provenance Initial Hessian scaling: gamma = (y'*s)/(y'*y) per Eq. 7.20
 */
export function lbfgs(
  f: (x: number[]) => number,
  x0: number[],
  grad?: (x: number[]) => number[],
  options?: Partial<LBFGSOptions>,
): OptimizeResult {
  const opts = defaultLBFGSOptions(options);
  const gradFn = grad ?? ((x: number[]) => forwardDiffGradient(f, x));

  let x = x0.slice();
  let fx = f(x);
  let gx = gradFn(x);
  let functionCalls = 1;
  let gradientCalls = 1;

  // Correction pair history (circular buffer, most recent at end)
  const sHistory: number[][] = [];
  const yHistory: number[][] = [];
  const rhoHistory: number[] = [];

  // Initial Hessian scaling
  let gamma = 1.0;

  // Check if already at a minimum
  let gradNorm = normInf(gx);
  const initialCheck = checkConvergence(gradNorm, Infinity, Infinity, 0, opts);
  if (initialCheck && isConverged(initialCheck)) {
    return {
      x, fun: fx, gradient: gx.slice(),
      iterations: 0, functionCalls, gradientCalls,
      converged: true, message: convergenceMessage(initialCheck),
    };
  }

  for (let iteration = 1; iteration <= opts.maxIterations; iteration++) {
    // Compute search direction using two-loop recursion
    let d: number[];
    if (sHistory.length === 0) {
      // First iteration: steepest descent
      d = negate(gx);
    } else {
      d = negate(twoLoopRecursion(gx, sHistory, yHistory, rhoHistory, gamma));
    }

    // Line search
    const ls = wolfeLineSearch(f, gradFn, x, d, fx, gx);
    functionCalls += ls.functionCalls;
    gradientCalls += ls.gradientCalls;

    if (!ls.success) {
      return {
        x, fun: fx, gradient: gx.slice(),
        iterations: iteration, functionCalls, gradientCalls,
        converged: false,
        message: "Stopped: line search failed to find acceptable step",
      };
    }

    const xNew = addScaled(x, d, ls.alpha);
    const fNew = ls.fNew;
    const gNew = ls.gNew ?? gradFn(xNew);
    if (!ls.gNew) gradientCalls++;

    const sk = sub(xNew, x);
    const yk = sub(gNew, gx);

    const stepNorm = normInf(sk);
    const funcChange = Math.abs(fNew - fx);
    gradNorm = normInf(gNew);

    x = xNew;
    fx = fNew;
    gx = gNew;

    // Check convergence
    const reason = checkConvergence(gradNorm, stepNorm, funcChange, iteration, opts);
    if (reason) {
      return {
        x: x.slice(), fun: fx, gradient: gx.slice(),
        iterations: iteration, functionCalls, gradientCalls,
        converged: isConverged(reason), message: convergenceMessage(reason),
      };
    }

    // Update history
    const ys = dot(yk, sk);
    if (ys > 1e-10) {
      if (sHistory.length >= opts.memory) {
        sHistory.shift();
        yHistory.shift();
        rhoHistory.shift();
      }
      sHistory.push(sk);
      yHistory.push(yk);
      rhoHistory.push(1.0 / ys);

      // Update initial Hessian scaling
      gamma = ys / dot(yk, yk);
    }
  }

  return {
    x: x.slice(), fun: fx, gradient: gx.slice(),
    iterations: opts.maxIterations, functionCalls, gradientCalls,
    converged: false,
    message: `Stopped: reached maximum iterations (${opts.maxIterations})`,
  };
}

File: research/experiments/sweep/runs/opt-full-csharp-sonnet45/src/LBfgs.cs
namespace Optimization;

public class LBfgsOptions : OptimizeOptions
{
    public int Memory { get; set; } = 10;
}

public static class LBfgs
{
    public static OptimizeResult Minimize(
        Func<double[], double> f, double[] x0,
        Func<double[], double[]>? grad = null, OptimizeOptions? options = null)
    {
        int memory = 10;
        if (options is LBfgsOptions lbOpts) memory = lbOpts.Memory;
        var opts = ResultTypes.MergeOptions(options);
        var gradFn = grad ?? (x => FiniteDiff.ForwardDiffGradient(f, x));
        int n = x0.Length;

        double[] x = (double[])x0.Clone();
        double fx = f(x);
        double[] gx = gradFn(x);
        int functionCalls = 1;
        int gradientCalls = 1;

        double gradNorm = VecOps.NormInf(gx);
        if (gradNorm <= opts.GradTol)
        {
            return new OptimizeResult
            {
                X = x, Fun = fx, Gradient = (double[])gx.Clone(),
                Iterations = 0, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
                Converged = true, Message = "Convergence: gradient norm below tolerance"
            };
        }

        var sHistory = new List<double[]>();
        var yHistory = new List<double[]>();
        var rhoHistory = new List<double>();
        double gammaK = 1.0;

        for (int iter = 1; iter <= opts.MaxIterations; iter++)
        {
            // Compute direction via two-loop recursion
            double[] d;
            if (sHistory.Count == 0)
            {
                d = VecOps.Negate(gx);
            }
            else
            {
                d = TwoLoopRecursion(gx, sHistory, yHistory, rhoHistory, gammaK);
            }

            var ls = LineSearch.WolfeLineSearch(f, gradFn, x, d, fx, gx);
            functionCalls += ls.FunctionCalls;
            gradientCalls += ls.GradientCalls;

            if (!ls.Success)
            {
                return new OptimizeResult
                {
                    X = (double[])x.Clone(), Fun = fx, Gradient = (double[])gx.Clone(),
                    Iterations = iter, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
                    Converged = false, Message = "Stopped: line search failed"
                };
            }

            double[] xNew = VecOps.AddScaled(x, d, ls.Alpha);
            double fNew = ls.FNew;
            double[] gNew = ls.GNew ?? gradFn(xNew);
            if (ls.GNew == null) gradientCalls++;

            double[] sk = VecOps.Sub(xNew, x);
            double[] yk = VecOps.Sub(gNew, gx);
            double yts = VecOps.Dot(yk, sk);

            double stepNorm = VecOps.NormInf(sk);
            double funcChange = Math.Abs(fNew - fx);

            x = xNew; fx = fNew; gx = gNew;
            gradNorm = VecOps.NormInf(gx);

            if (yts > 1e-10)
            {
                if (sHistory.Count >= memory)
                {
                    sHistory.RemoveAt(0);
                    yHistory.RemoveAt(0);
                    rhoHistory.RemoveAt(0);
                }
                sHistory.Add(sk);
                yHistory.Add(yk);
                rhoHistory.Add(1.0 / yts);
                gammaK = yts / VecOps.Dot(yk, yk);
            }

            var reason = ResultTypes.CheckConvergence(gradNorm, stepNorm, funcChange, iter, opts);
            if (reason != null)
            {
                return new OptimizeResult
                {
                    X = (double[])x.Clone(), Fun = fx, Gradient = (double[])gx.Clone(),
                    Iterations = iter, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
                    Converged = ResultTypes.IsConverged(reason), Message = ResultTypes.ConvergenceMessage(reason)
                };
            }
        }

        return new OptimizeResult
        {
            X = (double[])x.Clone(), Fun = fx, Gradient = (double[])gx.Clone(),
            Iterations = opts.MaxIterations, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
            Converged = false, Message = "Stopped: reached maximum iterations"
        };
    }

    private static double[] TwoLoopRecursion(
        double[] g, List<double[]> sHist, List<double[]> yHist, List<double> rhoHist, double gamma)
    {
        int n = g.Length;
        int m = sHist.Count;
        double[] q = (double[])g.Clone();
        double[] alphas = new double[m];

        for (int i = m - 1; i >= 0; i--)
        {
            alphas[i] = rhoHist[i] * VecOps.Dot(sHist[i], q);
            q = VecOps.AddScaled(q, yHist[i], -alphas[i]);
        }

        double[] r = VecOps.Scale(q, gamma);

        for (int i = 0; i < m; i++)
        {
            double beta = rhoHist[i] * VecOps.Dot(yHist[i], r);
            r = VecOps.AddScaled(r, sHist[i], alphas[i] - beta);
        }

        return VecOps.Negate(r);
    }
}
